optics-th/src/Optics/TH/Internal/Utils.hs

```diff
@@ -76,6 +78,48 @@ quantifyType' exclude vars cx t = ForallT vs cx t
     bndrToType (PlainTV n)    = VarT n
     bndrToType (KindedTV n k) = SigT (VarT n) k
 
+-- | Pass in a list of lists of extensions, where any of the given extensions
+-- will satisfy it. For example, you might need either GADTs or
+-- ExistentialQuantification, so you'd write:
+--
+-- > requireExtensions [[GADTs, ExistentialQuantification]]
+--
+-- But if you need TypeFamilies and MultiParamTypeClasses, then you'd write:
+--
+-- > requireExtensions [[TypeFamilies], [MultiParamTypeClasses]]
+--
+requireExtensions :: String -> [[Extension]] -> Q ()
+requireExtensions what extLists = do
+  -- Taken from the persistent library
+  required <- filterM (fmap (not . or) . traverse isExtEnabled) extLists
+  case mapMaybe listToMaybe required of
+    [] -> pure ()
+    [extension] -> fail $ mconcat
+      [ "Generating " ++ what ++ " requires the "
+      , show extension
+      , " language extension. Please enable it by copy/pasting this line to the top of your file:\n\n"
+      , extensionToPragma extension
```

Comment:
I like your addition below about the GHCi session. Might it be worth including for the single-extension case as well?

Reply:
Yeah, I forgot about that, done.